crates/orderbook/src/solvable_orders.rs

```diff
@@ -327,41 +334,64 @@ impl Maintaining for SolvableOrdersCache {
 async fn get_orders_with_native_prices(
     mut orders: Vec<Order>,
     native_price_estimator: &dyn NativePriceEstimating,
+    deadline: Instant,
 ) -> (Vec<Order>, BTreeMap<H160, U256>) {
     let traded_tokens = orders
         .iter()
         .flat_map(|order| [order.creation.sell_token, order.creation.buy_token])
         .collect::<HashSet<_>>()
         .into_iter()
         .collect::<Vec<_>>();
-    let prices = native_vec_estimates(native_price_estimator, &traded_tokens)
-        .await
-        .into_iter()
-        .zip(traded_tokens)
-        .filter_map(|(price, token)| match price {
-            Ok(price) => Some((token, to_normalized_price(price)?)),
-            Err(err) => {
-                tracing::warn!(?token, ?err, "error estimating native token price");
-                None
-            }
-        })
-        .collect::<BTreeMap<_, _>>();
+    let mut prices = BTreeMap::new();
+    let mut price_stream = native_price_estimator.estimate_native_prices(&traded_tokens);
+    let collect_prices = async {
+        while let Some((index, result)) = price_stream.next().await {
+            let token = &traded_tokens[index];
+            let price = match result {
+                Ok(price) => price,
+                Err(err) => {
+                    tracing::warn!(?token, ?err, "error estimating native token price");
+                    continue;
+                }
+            };
+            let price = match to_normalized_price(price) {
+                Some(price) => price,
+                None => continue,
+            };
+            prices.insert(*token, price);
+        }
+    };
+    match tokio::time::timeout_at(deadline, collect_prices).await {
+        Ok(()) => (),
+        Err(_) => tracing::warn!(
```

Comment:
Can we maybe add some metrics to this to?
- How many orders are missing?
- How often the timeout happens

Can be in a separate PR as well.